File: windows/Transkrito/AppController.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;
using System.Windows;
using System.Windows.Threading;
using Transkrito.Dictionary;
using Transkrito.Engine;
using Transkrito.History;
using Transkrito.Insert;
using Transkrito.Settings;
using Transkrito.Storage;

namespace Transkrito;

public enum RecordingState { Idle, Listening, Transcribing }

/// <summary>
/// The one object views bind to. Owns the recording state machine (shared/SPEC.md) and wires
/// hotkey → audio → engine → correction pass → history → clipboard/insert.
/// </summary>
public sealed class AppController : INotifyPropertyChanged, IDisposable
{
    public event PropertyChangedEventHandler? PropertyChanged;

    public DictionaryStore Dictionary { get; }
    public HistoryStore History { get; }
    public AppSettings Settings { get; }
    public AudioCapture Audio { get; } = new();

    /// <summary>The engine for the resolved model; swapped (and the old one released) when model or language changes.</summary>
    public ISpeechEngine Engine { get; private set; }

    private RecordingState _state = RecordingState.Idle;
    private string _status = "Ready";
    private bool _statusIsError;
    private double _level;
    private ModelInfo _model;
    private string? _hotkeyError;
    private DateTime _listeningSince;
    private string? _liveLanguage;          // confident guess from partial text while listening (Auto mode)
    private string? _lastDetected;          // last confident detection, used as a fallback hint
    private ModelInfo _loadedFor = null!;   // which resolved model Engine was created for

    public AppController()
    {
        AppPaths.EnsureDirs();
        Settings = AppSettings.Load();
        Dictionary = new DictionaryStore();
        History = new HistoryStore();
        _model = ModelCatalog.Get(Settings.Model);
        _loadedFor = EffectiveModel;
        Engine = _loadedFor.CreateEngine();
        Engine.PartialText += OnPartialText;
        Audio.DeviceId = Settings.InputDevice;
        Dictionary.Changed += () => Engine.SetBiasTerms(Bias.Terms(Dictionary.Entries));
        Audio.ChunkAvailable += chunk => { if (State == RecordingState.Listening) Engine.Feed(chunk); };
        Audio.LevelChanged += l => Application.Current?.Dispatcher.BeginInvoke(() => Level = l);
        Audio.Failed += ex => Application.Current?.Dispatcher.BeginInvoke(() => SetStatus($"Microphone error: {ex.Message}", error: true));
    }

    // ---- State ----

    public RecordingState State
    {
        get => _state;
        private set
        {
            if (_state == value) return;
            _state = value;
            OnChanged(); OnChanged(nameof(IsListening)); OnChanged(nameof(IsIdle)); OnChanged(nameof(IsTranscribing)); OnChanged(nameof(ActionLabel));
        }
    }
    public bool IsListening => State == RecordingState.Listening;
    public bool IsTranscribing => State == RecordingState.Transcribing;
    public bool IsIdle => State == RecordingState.Idle;
    public string ActionLabel => State == RecordingState.Listening ? "Stop" : "Start";

    public string Status { get => _status; private set { _status = value; OnChanged(); } }
    public bool StatusIsError { get => _statusIsError; private set { _statusIsError = value; OnChanged(); } }
    public double Level { get => _level; private set { _level = value; OnChanged(); } }

    public void SetStatus(string text, bool error = false) { Status = text; StatusIsError = error; }

    // ---- Settings-facing ----

    /// <summary>The user's model pick. The model that actually runs is <see cref="EffectiveModel"/>.</summary>
    public ModelInfo Model
    {
        get => _model;
        set { _model = value; Settings.Model = value.Id; Settings.Save(); OnChanged(); OnChanged(nameof(EffectiveModel)); OnChanged(nameof(ModelCompatibility)); _ = LoadModelAsync(); }
    }

    /// <summary>Spoken language setting: auto | en | de | ar. Takes effect on the next recording; no restart.</summary>
    public string Language
    {
        get => Settings.Language;
        set
        {
            if (Settings.Language == value) return;
            Settings.Language = value; Settings.Save();
            OnChanged(); OnChanged(nameof(EffectiveModel)); OnChanged(nameof(ModelCompatibility)); OnChanged(nameof(LanguageHint));
            _ = LoadModelAsync();
        }
    }
    public string LanguageHint => Language == Lang.Auto
        ? "Automatically detects English, German, or Arabic."
        : "Always transcribe using this language.";

    /// <summary>Model resolved from the pick and the language: Parakeet cannot do Arabic or Auto, so those route to Nemotron.</summary>
    public ModelInfo EffectiveModel => ModelCatalog.Resolve(Settings.Model, Settings.Language);
    public string ModelCompatibility => EffectiveModel.Id == Model.Id
        ? ""
        : $"{Model.Name} cannot transcribe {(Language == Lang.Auto ? "in Auto mode" : Lang.Display(Language))} — {EffectiveModel.Name} is used instead.";

    public string HotkeyLabel => Settings.Hotkey.Display();
    public string HotkeyHint => Settings.HoldToTalk ? "Hold to dictate" : "Press to start, press to stop";
    public string HoldToTalkVerb => Settings.HoldToTalk ? "Hold" : "Press";
    public bool HoldToTalk
    {
        get => Settings.HoldToTalk;
        set { Settings.HotkeyMode = value ? "hold" : "toggle"; Settings.Save(); OnChanged(); OnChanged(nameof(HotkeyHint)); OnChanged(nameof(HoldToTalkVerb)); }
    }

    /// <summary>Set when the keyboard hook could not be installed. Shown in place of the hotkey hint.</summary>
    public string? HotkeyError
    {
        get => _hotkeyError;
        set { _hotkeyError = value; OnChanged(); OnChanged(nameof(HotkeyHasError)); }
    }
    public bool HotkeyHasError => HotkeyError is not null;
    public void HotkeyChanged() { OnChanged(nameof(HotkeyLabel)); }

    public string InputDeviceId
    {
        get => Settings.InputDevice;
        set { Settings.InputDevice = value; Settings.Save(); Audio.DeviceId = value; OnChanged(); }
    }

    // ---- Model ----

    /// <summary>Called once at startup and whenever the model pick or the language changes.</summary>
    public async Task LoadModelAsync()
    {
        var target = EffectiveModel;
        if (target.Id != _loadedFor.Id)
        {
            // Release the previous engine so only one model stays resident.
            Engine.PartialText -= OnPartialText;
            Engine.Dispose();
            Engine = target.CreateEngine();
            Engine.PartialText += OnPartialText;
            _loadedFor = target;
            OnChanged(nameof(Engine));
        }
        if (!target.IsInstalled)
        {
            SetStatus($"{target.Name} not downloaded — open Settings", error: true);
            return;
        }
        if (Engine.IsLoaded && Engine.LoadedModel?.Id == target.Id) { SetStatus("Ready"); return; }
        SetStatus("Loading model…");
        try
        {
            await Engine.LoadAsync(target);
            Engine.SetBiasTerms(Bias.Terms(Dictionary.Entries));
            SetStatus("Ready");
        }
        catch (Exception ex)
        {
            SetStatus($"Model failed to load: {ex.Message}", error: true);
        }
    }

    /// <summary>Streaming partial text → live language for the status line ("Listening · Deutsch").</summary>
    private void OnPartialText(string partial)
    {
        if (Settings.Language != Lang.Auto) return;
        var guess = LanguageDetector.Detect(partial);
        if (!guess.IsConfident || guess.Language == _liveLanguage) return;
        _liveLanguage = guess.Language;
        Application.Current?.Dispatcher.BeginInvoke(() =>
        {
            if (State == RecordingState.Listening) SetStatus($"Listening · {Lang.Display(_liveLanguage)}");
        });
    }

    // ---- Hotkey entry points ----

    /// <summary>Chord pressed. Hold mode: start. Toggle mode: start or stop.</summary>
    public void HotkeyDown()
    {
        if (Settings.HoldToTalk) Start();
        else Toggle();
    }

    /// <summary>Chord released. Hold mode: stop (a tap shorter than the guard is treated as a mis-press).</summary>
    public void HotkeyUp()
    {
        if (!Settings.HoldToTalk) return;
        if (State != RecordingState.Listening) return;
        _ = StopAsync();
    }

    public void Toggle()
    {
        switch (State)
        {
            case RecordingState.Idle: Start(); break;
            case RecordingState.Listening: _ = StopAsync(); break;
        }
    }

    public void Start()
    {
        if (State != RecordingState.Idle) return;
        if (!Engine.IsLoaded)
        {
            SetStatus(EffectiveModel.IsInstalled ? "Model is still loading" : $"{EffectiveModel.Name} not downloaded — open Settings", error: true);
            return;
        }
        try
        {
            _liveLanguage = null;
            Engine.BeginUtterance(Settings.Language == Lang.Auto ? null : Settings.Language);
            Audio.Start();
            _listeningSince = DateTime.UtcNow;
            State = RecordingState.Listening;
            SetStatus(Settings.Language == Lang.Auto ? "Listening" : $"Listening · {Lang.Display(Settings.Language)}");
        }
        catch (Exception ex)
        {
            SetStatus($"Microphone unavailable: {ex.Message}", error: true);
        }
    }

    public async Task StopAsync()
    {
        if (State != RecordingState.Listening) return;
        var (_, duration) = Audio.Stop();
        State = RecordingState.Transcribing;
        SetStatus("Transcribing…");
        try
        {
            var raw = await Engine.EndUtteranceAsync();
            if (string.IsNullOrWhiteSpace(raw))
            {
                SetStatus(duration < 0.4 ? "Hold the key while you speak" : "Nothing heard");
                return;
            }
            var (text, events) = CorrectionEngine.Apply(raw, Dictionary.Entries);
            // Explicit mode: the language is what the user pinned. Auto: what the transcript's script/lexicon says, else unknown.
            var language = Settings.Language == Lang.Auto ? LanguageDetector.ForHistory(raw) : Settings.Language;
            if (language != Lang.Unknown) _lastDetected = language;
            var item = new Transcription
            {
                Raw = raw,
                Text = text,
                DurationSec = Math.Round(duration, 2),
                Engine = EffectiveModel.Kind == EngineKind.NemotronStreaming ? "nemotron" : "parakeet",
                Model = EffectiveModel.Id,
                Language = language,
                Corrections = events.Select(CorrectionRecord.From).ToList(),
            };
            History.Add(item);
            ClearNewLater(item);
            Inserter.CopyToClipboard(text);
            if (Settings.InsertAtCursor) Inserter.PasteIntoForegroundApp();
            SetStatus(events.Count == 0 ? "Copied" : $"Copied · {item.CorrectionLabel}");
        }
        catch (Exception ex)
        {
            SetStatus($"Transcription failed: {ex.Message}", error: true);
        }
        finally
        {
            State = RecordingState.Idle;
            Level = 0;
        }
    }

    /// <summary>New rows stay tinted until noticed (motion.rowHighlight), then settle.</summary>
    private static void ClearNewLater(Transcription t)
    {
        var timer = new DispatcherTimer { Interval = TimeSpan.FromMilliseconds(Design.Tokens.Motion.RowHighlightMs) };
        timer.Tick += (_, _) => { t.IsNew = false; timer.Stop(); };
        timer.Start();
    }

    private void OnChanged([CallerMemberName] string? name = null) =>
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));

    public void Dispose()
    {
        Audio.Dispose();
        Engine.Dispose();
        Dictionary.Dispose();
    }
}

File: macos/Sources/Transkrito/App/AppController.swift
import AppKit
import AVFoundation
import Foundation
import Observation

enum RecordingState { case idle, listening, transcribing }

/// The one object views observe. Owns the recording state machine (shared/SPEC.md) and wires
/// audio → (language detection) → engine → correction pass → history → clipboard/insert.
///
/// Languages: en/de/ar explicit, or Auto. Explicit modes open the Apple session for that locale straight away.
/// Auto buffers the first second or so of speech, runs the local Whisper-tiny language ID on it, opens the matching
/// pipeline and replays the buffered audio into it — nothing is lost. Arabic uses Apple Speech when this Mac can
/// install its Arabic asset, otherwise the local Nemotron multilingual model (same as the Windows build).
@MainActor @Observable
final class AppController {
    let dictionary: DictionaryStore
    let history: HistoryStore
    var settings: AppSettings
    let engine: SpeechEngine
    let audio = AudioCapture()
    let hotkey = HotkeyManager()
    let languageID = SpokenLanguageID()
    let nemotron = NemotronEngine()

    private(set) var state: RecordingState = .idle
    private(set) var status = "Ready"
    private(set) var statusIsError = false
    private(set) var level: Double = 0
    private(set) var hotkeyError: String?
    /// Sherpa model download progress (0…1, -1 while unpacking) or nil.
    private(set) var sherpaProgress: [String: Double] = [:]
    /// True once Apple Speech proved unable to provide Arabic on this Mac; the Nemotron path is used instead.
    private(set) var arabicFallback = false
    private var arabicFallbackAnnounced = false

    var isListening: Bool { state == .listening }
    var isTranscribing: Bool { state == .transcribing }
    var isIdle: Bool { state == .idle }
    var actionLabel: String { state == .listening ? "Stop" : "Start" }
    var hotkeyLabel: String { settings.hotkey.display }
    var hotkeyHint: String { settings.holdToTalk ? "Hold to dictate" : "Press to start, press to stop" }
    var holdToTalkVerb: String { settings.holdToTalk ? "Hold" : "Press" }
    var holdToTalk: Bool {
        get { settings.holdToTalk }
        set { settings.hotkeyMode = newValue ? "hold" : "toggle"; settings.save() }
    }
    var inputDevice: String {
        get { settings.inputDevice }
        set { settings.inputDevice = newValue; settings.save(); audio.deviceUID = newValue }
    }
    var language: String {
        get { settings.language }
        set {
            guard settings.language != newValue else { return }
            settings.language = newValue; settings.save()
            Task { await loadModel() }
        }
    }
    var languageHint: String {
        settings.language == Lang.auto ? "Automatically detects English, German, or Arabic." : "Always transcribe using this language."
    }

    // MARK: - Utterance pipeline

    private enum Phase { case idle, detecting, apple, nemotron }
    private final class Utterance: @unchecked Sendable {
        let lock = NSLock()
        var phase: Phase = .detecting
        var buffers: [AVAudioPCMBuffer] = []   // preroll in the mic's format (for Apple)
        var pcm: [Float] = []                  // preroll at 16 kHz mono (for language ID / Nemotron)
        var voicedSeconds = 0.0
        var totalSeconds = 0.0
        var decisionRequested = false
        var language: String?                  // decided language
        var uncertain = false
        var inputFormat: AVAudioFormat?
        let converter = PCM16k()
    }
    /// Read from the audio thread, replaced only on the main actor at start/stop.
    private final class UtteranceBox: @unchecked Sendable {
        private let lock = NSLock()
        private var value: Utterance?
        var current: Utterance? {
            get { lock.lock(); defer { lock.unlock() }; return value }
            set { lock.lock(); value = newValue; lock.unlock() }
        }
    }
    private let utteranceBox = UtteranceBox()
    private var utterance: Utterance? {
        get { utteranceBox.current }
        set { utteranceBox.current = newValue }
    }
    private var lastDetected: (language: String, at: Date)?

    init() {
        AppPaths.ensureDirs()
        settings = AppSettings.load()
        dictionary = DictionaryStore()
        history = HistoryStore()
        engine = SpeechEngine(localeIdentifier: settings.model)
        engine.biasTerms = Bias.terms(dictionary.entries)
        dictionary.onChange = { [weak self] in
            guard let self else { return }
            self.engine.biasTerms = Bias.terms(self.dictionary.entries)
        }
        audio.deviceUID = settings.inputDevice
        audio.onLevel = { [weak self] l in Task { @MainActor in self?.level = l } }
        audio.onBuffer = { [weak self] b in self?.onBuffer(b) }
        nemotron.onPartial = { [weak self] text in Task { @MainActor in self?.onPartial(text) } }
        hotkey.onPressed = { [weak self] in self?.hotkeyDown() }
        hotkey.onReleased = { [weak self] in self?.hotkeyUp() }
        applyHotkey(settings.hotkey)
    }

    func setStatus(_ text: String, error: Bool = false) { status = text; statusIsError = error }

    // MARK: - Settings

    func applyHotkey(_ hk: HotkeySetting) {
        settings.hotkey = hk
        settings.save()
        hotkeyError = hotkey.register(hk) ? nil : hotkey.lastError
    }

    func setInsertAtCursor(_ on: Bool) {
        settings.insertAtCursor = on
        settings.save()
    }

    // MARK: - Models

    /// Makes sure everything the current language setting needs is present: Apple assets (installing on demand),
    /// the language detector for Auto, and the Nemotron fallback when Apple cannot do Arabic here.
    func loadModel() async {
        await engine.refreshAllAssetStates()
        let needed = settings.language == Lang.auto ? Lang.supported : [settings.language]
        for l in needed where l != Lang.ar {
            if engine.assetState(for: l) == .notInstalled {
                setStatus("Downloading \(Lang.display(l)) speech model\u{2026}")
                await engine.installAsset(for: l)
            }
        }
        if needed.contains(Lang.ar) { await ensureArabic() }
        if settings.language == Lang.auto { await ensureLanguageID() }

        let missing = needed.filter { l in l == Lang.ar ? !(engine.isReady(for: Lang.ar) || (arabicFallback && nemotron.isLoaded)) : !engine.isReady(for: l) }
        if !missing.isEmpty {
            setStatus("\(missing.map(Lang.display).joined(separator: ", ")) speech model unavailable \u{2014} open Settings", error: true)
        } else if settings.language == Lang.auto && !languageID.isLoaded {
            setStatus("Language detector unavailable \u{2014} open Settings", error: true)
        } else if arabicFallback && !arabicFallbackAnnounced && needed.contains(Lang.ar) {
            arabicFallbackAnnounced = true
            setStatus("Apple Arabic speech model unavailable \u{2014} using local multilingual model")
        } else {
            setStatus("Ready")
        }
    }

    /// Arabic is tested at runtime: resolve a locale, install its asset, and only fall back when that truly fails.
    private func ensureArabic() async {
        if engine.isReady(for: Lang.ar) { arabicFallback = false; return }
        if engine.assetState(for: Lang.ar) == .notInstalled {
            setStatus("Downloading Arabic speech model\u{2026}")
            if await engine.installAsset(for: Lang.ar) { arabicFallback = false; return }
        }
        arabicFallback = true
        if !nemotron.isLoaded {
            if !SherpaCatalog.nemotron.isInstalled { await downloadSherpa(SherpaCatalog.nemotron, label: "Arabic speech model") }
            if SherpaCatalog.nemotron.isInstalled {
                setStatus("Loading Arabic speech model\u{2026}")
                do { try await Task.detached { [nemotron] in try nemotron.load() }.value } catch { setStatus("Arabic model unavailable", error: true) }
            }
        }
    }

    private func ensureLanguageID() async {
        if languageID.isLoaded { return }
        if !SherpaCatalog.languageID.isInstalled { await downloadSherpa(SherpaCatalog.languageID, label: "language detector") }
        if SherpaCatalog.languageID.isInstalled {
            do { try await Task.detached { [languageID] in try languageID.load() }.value } catch { setStatus("Language detector unavailable", error: true) }
        }
    }

    func downloadSherpa(_ model: SherpaModel, label: String) async {
        setStatus("Downloading \(label)\u{2026}")
        sherpaProgress[model.id] = 0
        do {
            try await SherpaModelManager.download(model) { [weak self] p in Task { @MainActor in self?.sherpaProgress[model.id] = p } }
        } catch {
            setStatus("\(label.capitalized) download failed: \(error.localizedDescription)", error: true)
        }
        sherpaProgress[model.id] = nil
    }

    // MARK: - Recording

    /// Chord pressed. Hold mode: start. Toggle mode: start or stop.
    func hotkeyDown() {
        if settings.holdToTalk { Task { await start() } } else { toggle() }
    }

    /// Chord released. Hold mode: stop.
    func hotkeyUp() {
        guard settings.holdToTalk, state == .listening else { return }
        Task { await stop() }
    }

    func toggle() {
        switch state {
        case .idle: Task { await start() }
        case .listening: Task { await stop() }
        case .transcribing: break
        }
    }

    func start() async {
        guard state == .idle else { return }
        guard await AudioCapture.requestPermission() else {
            setStatus("Microphone access denied. Allow Transkrito in System Settings \u{2192} Privacy & Security \u{2192} Microphone.", error: true)
            return
        }
        let u = Utterance()
        let format = audio.inputFormat
        u.inputFormat = format
        utterance = u
        do {
            if settings.language == Lang.auto {
                if !languageID.isLoaded { await loadModel() }
                guard languageID.isLoaded else { utterance = nil; return }
                u.phase = .detecting
                try audio.start()
                state = .listening
                setStatus("Detecting language\u{2026}")
            } else {
                try await openPipeline(u, language: settings.language, uncertain: false)
                try audio.start()
                state = .listening
                setStatus("Listening \u{00B7} \(Lang.display(settings.language))")
            }
        } catch {
            utterance = nil
            await engine.cancelSession()
            setStatus("Could not start: \(error.localizedDescription)", error: true)
        }
    }

    /// Audio thread. Preroll while detecting; otherwise feed the active pipeline.
    nonisolated private func onBuffer(_ buffer: AVAudioPCMBuffer) {
        guard let u = utteranceBox.current else { return }
        u.lock.lock()
        switch u.phase {
        case .detecting:
            let pcm = u.converter.convert(buffer)
            u.buffers.append(buffer)
            u.pcm.append(contentsOf: pcm)
            let dt = Double(buffer.frameLength) / buffer.format.sampleRate
            u.totalSeconds += dt
            if audio.currentLevel >= Tokens.Pillar.speakingThreshold { u.voicedSeconds += dt }
            let ready = u.voicedSeconds >= AutoDetect.minVoicedSeconds || u.totalSeconds >= AutoDetect.maxWaitSeconds
            let request = ready && !u.decisionRequested
            if request { u.decisionRequested = true }
            u.lock.unlock()
            if request { Task { @MainActor in await self.decide() } }
        case .apple:
            u.lock.unlock()
            engine.feed(buffer)
        case .nemotron:
            let pcm = u.converter.convert(buffer)
            u.lock.unlock()
            nemotron.feed(pcm)
        case .idle:
            u.lock.unlock()
        }
    }

    /// Runs the language ID on the preroll, opens the pipeline, replays the preroll into it.
    private func decide() async {
        guard let u = utterance else { return }
        u.lock.lock()
        guard u.phase == .detecting else { u.lock.unlock(); return }
        let sample = u.pcm
        u.lock.unlock()

        let detected = await Task.detached { [languageID] in languageID.detect(sample) }.value
        let (lang, uncertain) = detected.map { ($0, false) } ?? fallbackLanguage()
        do {
            try await openPipeline(u, language: lang, uncertain: uncertain)
        } catch {
            await engine.cancelSession()
            setStatus("Could not start: \(error.localizedDescription)", error: true)
            return
        }
        if state == .listening { setStatus("Listening \u{00B7} \(Lang.display(lang))") }
    }

    /// Opens Apple (or Nemotron for Arabic-without-Apple) for `language`, then flushes any preroll — under the lock so
    /// the audio thread cannot slip a buffer past the switch.
    private func openPipeline(_ u: Utterance, language: String, uncertain: Bool) async throws {
        let useNemotron = language == Lang.ar && (arabicFallback || !engine.isReady(for: Lang.ar))
        if useNemotron {
            guard nemotron.isLoaded else { throw SpeechEngine.EngineError.languageUnsupported(Lang.ar) }
            nemotron.begin()
        } else {
            guard let format = u.inputFormat else { return }
            try await engine.startSession(inputFormat: format, language: language)
        }
        u.lock.lock()
        let buffers = u.buffers, pcm = u.pcm
        u.buffers = []; u.pcm = []
        u.language = language; u.uncertain = uncertain
        u.phase = useNemotron ? .nemotron : .apple
        u.lock.unlock()
        if useNemotron { if !pcm.isEmpty { nemotron.feed(pcm) } } else { for b in buffers { engine.feed(b) } }
    }

    /// When the detector is unsure: the last recent detection, else the system language if exposed, else English.
    private func fallbackLanguage() -> (String, Bool) {
        if let last = lastDetected, Date().timeIntervalSince(last.at) < AutoDetect.recentSeconds { return (last.language, true) }
        if let sys = Locale.current.language.languageCode?.identifier, Lang.isSupported(sys) { return (sys, true) }
        return (Lang.en, true)
    }

    private func onPartial(_ text: String) { /* Nemotron partials: nothing to show beyond the language line */ }

    func stop() async {
        guard state == .listening, let u = utterance else { return }
        let duration = audio.stop()
        state = .transcribing
        setStatus("Transcribing\u{2026}")
        defer { state = .idle; level = 0; utterance = nil }
        do {
            // Released before the detector had enough audio: decide on what we have.
            u.lock.lock(); let stillDetecting = u.phase == .detecting; u.lock.unlock()
            if stillDetecting { await decide() }
            u.lock.lock(); let phase = u.phase; let chosen = u.language; let uncertain = u.uncertain; u.lock.unlock()

            let raw: String
            switch phase {
            case .apple: raw = try await engine.finishSession()
            case .nemotron: raw = await nemotron.end()
            default: raw = ""
            }
            if raw.isEmpty {
                setStatus(duration < 0.4 ? "Hold the key while you speak" : "Nothing heard")
                return
            }
            let (text, events) = CorrectionEngine.apply(raw, entries: dictionary.entries)
            let language: String
            if settings.language != Lang.auto {
                language = settings.language
            } else {
                let textGuess = LanguageDetector.detect(raw)
                if let chosen, !uncertain { language = chosen }
                else if textGuess.isConfident { language = textGuess.language }
                else { language = Lang.unknown }
            }
            if language != Lang.unknown { lastDetected = (language, Date()) }
            let item = Transcription(
                raw: raw, text: text, durationSec: (duration * 100).rounded() / 100,
                engine: phase == .nemotron ? "nemotron" : "apple-speech",
                model: phase == .nemotron ? SherpaCatalog.nemotron.id : engine.locale.identifier,
                corrections: events.map(CorrectionRecord.init),
                language: language
            )
            history.add(item)
            let id = item.id
            Task { try? await Task.sleep(for: .seconds(Tokens.Motion.rowHighlight)); history.markSeen(id) }
            Inserter.copyToClipboard(text)
            if settings.insertAtCursor { Inserter.pasteIntoForegroundApp() }
            setStatus(events.isEmpty ? "Copied" : "Copied \u{00B7} \(item.correctionLabel)")
        } catch {
            setStatus("Transcription failed: \(error.localizedDescription)", error: true)
        }
    }
}

/// Auto-detection timing (seconds). Enough voiced audio for a meaningful decision, bounded so silence cannot stall.
enum AutoDetect {
    static let minVoicedSeconds = 1.2
    static let maxWaitSeconds = 3.0
    static let recentSeconds = 600.0
}
